File: portfolio-pnl-simulator/src/analytics/table.py
import numpy as np
import pandas as pd

from src.core.utils import get_norm_strikes
from src.core.models.market import Market
from src.core.models.option import Option
from src.core.models.optiontype import OptionType
from src.core.workers.calculator import calculate_risk, get_forward_price, get_forward_strike, price_option

def get_option_chain_table(market: Market, expiry: float) -> pd.DataFrame:
    strikes = []
    fwd_strikes = []
    call_prices = []
    call_deltas = []
    call_gammas = []
    call_vegas = []
    call_thetas = []
    put_prices = []
    put_deltas = []
    put_gammas = []
    put_vegas = []
    put_thetas = []
    ivs = []
    forward = get_forward_price(market.spot, market.interest_rate, expiry)
    norm_strikes = get_norm_strikes(market, expiry, market.vol_surface.get_atm_vol(expiry))
    for ns in norm_strikes:
        strike = market.vol_surface.get_strike(ns, forward, expiry)
        option_call = Option(
            asset=market.asset,
            strike=strike,
            expiry=expiry,
            option_type=OptionType.CALL,
            lot_size=1
        )
        option_put = Option(
            asset=market.asset,
            strike=strike,
            expiry=expiry,
            option_type=OptionType.PUT,
            lot_size=1
        )
        call_price = price_option(option_call, market)
        call_risk = calculate_risk(option_call, market)
        put_price = price_option(option_put, market)
        put_risk = calculate_risk(option_put, market)

        strikes.append(strike)
        fwd_strikes.append(get_forward_strike(market, strike, expiry))
        ivs.append(market.vol_surface.get_vol_ns(ns, expiry))
        call_prices.append(call_price)
        call_deltas.append(call_risk.delta)
        call_gammas.append(call_risk.gamma)
        call_vegas.append(call_risk.vega / 100) # per 1% vol move
        call_thetas.append(call_risk.theta / 365) # per day
        put_prices.append(put_price)
        put_deltas.append(put_risk.delta)
        put_gammas.append(put_risk.gamma)
        put_vegas.append(put_risk.vega / 100) # per 1% vol move
        put_thetas.append(put_risk.theta / 365) # per day

    df = pd.DataFrame({
        'Call Theta': call_thetas,
        'Call Vega': call_vegas,
        'Call Gamma': call_gammas,
        'Call Delta': call_deltas,
        'Call Price': call_prices,
        'Norm Strike': norm_strikes,
        'Strike': strikes,
        'Forward Strike': fwd_strikes,
        'Implied Vol': ivs,
        'Put Price': put_prices,
        'Put Delta': put_deltas,
        'Put Gamma': put_gammas,
        'Put Vega': put_vegas,
        'Put Theta': put_thetas
    })
    return df

File: options-market-making-temp/src/analytics/table.cpp
#include "analytics/table.hpp"
#include "core/utils.hpp"
#include "core/workers/calculator.hpp"
#include <iostream>
#include <iomanip>
#include <vector>
#include <algorithm>

namespace omm::analytics {

std::vector<OptionChainRow> Table::getOptionChainTable(
    const std::shared_ptr<omm::core::models::Market>& market,
    double expiry
) {
    std::vector<OptionChainRow> rows;
    
    double forward = omm::core::Utils::getForwardPrice(market->spot, market->interestRate, expiry);
    std::vector<double> normStrikes = omm::core::Utils::getNormStrikes(
        market->spot,
        market->interestRate,
        expiry,
        market->volSurface->getAtmVol(expiry)
    );
    
    using namespace omm::core::models;
    using namespace omm::core::workers;
    
    for (double ns : normStrikes) {
        double strike = market->volSurface->getStrike(ns, forward, expiry);
        
        auto optionCall = std::make_shared<Option>(
            market->asset,
            strike,
            expiry,
            OptionType::CALL,
            1
        );
        
        auto optionPut = std::make_shared<Option>(
            market->asset,
            strike,
            expiry,
            OptionType::PUT,
            1
        );
        
        double callPrice = Calculator::priceOption(*optionCall, *market);
        Risk callRisk = Calculator::calculateRisk(*optionCall, *market);
        double putPrice = Calculator::priceOption(*optionPut, *market);
        Risk putRisk = Calculator::calculateRisk(*optionPut, *market);
        
        double forwardStrike = strike * std::exp(market->interestRate * expiry / 365.0);
        double impliedVol = market->volSurface->getVolNormStrike(ns, expiry);
        
        OptionChainRow row{
            callRisk.theta / 365.0,           // callTheta per day
            callRisk.vega / 100.0,            // callVega per 1% vol move
            callRisk.gamma,                   // callGamma
            callRisk.delta,                   // callDelta
            callPrice,                        // callPrice
            ns,                               // normStrike
            strike,                           // strike
            forwardStrike,                    // forwardStrike
            impliedVol,                       // impliedVol
            putPrice,                         // putPrice
            putRisk.delta,                    // putDelta
            putRisk.gamma,                    // putGamma
            putRisk.vega / 100.0,             // putVega per 1% vol move
            putRisk.theta / 365.0             // putTheta per day
        };
        
        rows.push_back(row);
    }
    
    return rows;
}

void Table::printOptionChainTable(const std::vector<OptionChainRow>& chain) {
    if (chain.empty()) {
        std::cout << "Empty option chain table\n";
        return;
    }
    
    std::cout << std::fixed << std::setprecision(4);
    std::cout << std::string(200, '=') << "\n";
    std::cout << std::setw(12) << "Call Theta"
              << std::setw(12) << "Call Vega"
              << std::setw(12) << "Call Gamma"
              << std::setw(12) << "Call Delta"
              << std::setw(12) << "Call Price"
              << std::setw(12) << "Norm Strike"
              << std::setw(12) << "Strike"
              << std::setw(12) << "Fwd Strike"
              << std::setw(12) << "Impl Vol"
              << std::setw(12) << "Put Price"
              << std::setw(12) << "Put Delta"
              << std::setw(12) << "Put Gamma"
              << std::setw(12) << "Put Vega"
              << std::setw(12) << "Put Theta\n";
    std::cout << std::string(200, '-') << "\n";
    
    for (const auto& row : chain) {
        std::cout << std::setw(12) << row.callTheta
                  << std::setw(12) << row.callVega
                  << std::setw(12) << row.callGamma
                  << std::setw(12) << row.callDelta
                  << std::setw(12) << row.callPrice
                  << std::setw(12) << row.normStrike
                  << std::setw(12) << row.strike
                  << std::setw(12) << row.forwardStrike
                  << std::setw(12) << row.impliedVol
                  << std::setw(12) << row.putPrice
                  << std::setw(12) << row.putDelta
                  << std::setw(12) << row.putGamma
                  << std::setw(12) << row.putVega
                  << std::setw(12) << row.putTheta << "\n";
    }
    std::cout << std::string(200, '=') << "\n";
}

}  // namespace omm::analytics
